src/bepasty/tests/test_rest_server.py

```diff
@@ -0,0 +1,817 @@
+        if now and time.time != self.get_time:
+            time.time = self.get_time
+        elif now is None and time.time == self.get_time:
+            time.time = self.orig
+
+    def get_time(self):
+        return self.now
+
+    def __enter__(self):
+        self.set_time(self.now)
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.set_time(None)
+
+
+def wait_background():
+    """Wait until background threads terminate."""
+
+    main_thread = threading.current_thread()
+    for t in threading.enumerate():
+        if t is not main_thread:
+            t.join()
+
+
+@pytest.fixture
+def client_fixture(tmp_path):
+    with FakeTime() as faketime:
+        Config.PERMISSIONS = {
+            'admin': 'admin,list,create,read,delete',
+            'full': 'list,create,read,delete',
+            'none': '',
+        }
+        Config.STORAGE_FILESYSTEM_DIRECTORY = str(tmp_path)
+
+        app = create_app()
+        app.config['TESTING'] = True
+
+        with app.test_client() as client:
+            with app.app_context():
+                yield app, client, faketime
+
+        wait_background()
+
+
+class TmpConfig:
+    def __init__(self, app, tmp_config=None):
+        self.config = app.config
+        self.orig_config = copy.deepcopy(app.config)
+        self.tmp_config = {} if tmp_config is None else tmp_config
+
+    def __enter__(self):
+        self.config.update(self.tmp_config)
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.config.update(self.orig_config)
+
+
+class RestUrl:
+    def __init__(self, fid=None):
+        self.fid = fid
+
+    @property
+    def config(self):
+        with current_app.test_request_context():
+            return url_for('bepasty_apis.api_info')
+
+    @property
+    def upload(self):
+        with current_app.test_request_context():
+            return url_for('bepasty_apis.items')
+
+    @property
+    def list(self):
+        return self.upload
+
+    @property
+    def detail(self):
+        with current_app.test_request_context():
+            return url_for('bepasty_apis.items_detail', name=self.fid)
+
+    @property
+    def download(self):
+        with current_app.test_request_context():
+            return url_for('bepasty_apis.items_download', name=self.fid)
+
+
+def add_auth(user, password, headers=None):
+    headers = headers if headers is not None else {}
+    headers['Authorization'] = _basic_auth_str(user, password)
+    return headers
+
+
+def test_invalid_url(client_fixture):
+    _, client, _ = client_fixture
+
+    response = client.get('/apis/rest/invalid')
+    assert response.status_code == 404
+
+
+def check_response(response, code, ftype='application/json', check_data=True):
+    assert response.status_code == code
+    assert response.headers['Content-Type'] == ftype
+    if check_data:
+        assert int(response.headers['Content-Length']) == len(response.data)
+
+
+re_html_tag = re.compile(r'<.+>')
```

Comment:
if this is only used in the function below, move it into that scope. we don't need to micro-optimize here.

Reply:
changed